### PR TITLE
[Backport] Fix for #16069: Configurable product price is not displayed if all children are out of stock and even if Display Out of Stock Products is set to "yes" #18776

### PR DESCRIPTION
**Pull request by vpodorozh**
### Original Pull Request
https://github.com/magento/magento2/pull/18776

### Description (#16069)
Configurable product price is not displayed if all children are out of stock and even if Display Out of Stock Products is set to "yes"

### Fixed Issues
1. magento/magento2#16069: Configurable product price is not displayed if all children are out of stock and even if Display Out of Stock Products is set to "yes"

### Preconditions
1. Magento CE 2.2.2 with Sample Data
2. PHP 7.0

### Manual testing scenarios
1. Create configurable product with view simples
2. Make configurable and all simple products out of stock
3. Set Display Out of Stock Products to "yes"

#### Expected result
Price for configurable product should be displayed

#### Actual result
Price is not displayed at all

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by vpodorozh** [comment]
@novikor - please review

File: app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/StockStatus.php [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        $isAllChildOutOfStock = true;
+        foreach ($statuses as $status) {
+            if ($status === CatalogInventoryStockStatusInterface::STATUS_IN_STOCK) {
+                $isAllChildOutOfStock = false;
+                break;
+            }
+        }
+
+        $this->allChildOutOfStockInfo[$productId] = $isAllChildOutOfStock;
+        return $this->allChildOutOfStockInfo[$productId];
```
Comment (novikor) [inline app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/StockStatus.php:64]: Please a blank line before return statement

**Review by novikor — CHANGES_REQUESTED** [review]
Please add a blank line (comment is left near needed piece of code).

File: app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/StockStatus.php (hunk 1 above)
Comment (vpodorozh) [reply app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/StockStatus.php:64]: done

Comment (vpodorozh) [reply app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/StockStatus.php:64]: Done

Comment (orlangur) [reply app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/StockStatus.php:64]: @novikor not needed, please concentrate on crucial things during review.

File: app/code/Magento/Catalog/Pricing/Render/FinalPriceBox.php [hunk 2]
```diff
@@ -86,6 +88,15 @@ protected function _toHtml()
         return $this->wrapResult($result);
     }
 
+    /**
+     * @param SaleableInterface $salableItem
+     * @return bool
+     */
+    protected function isApplySalableCheck(SaleableInterface $salableItem): bool
+    {
+        return (bool)$salableItem;
```
Comment (orlangur) [inline app/code/Magento/Catalog/Pricing/Render/FinalPriceBox.php]: Object will be always converted to `true`, isn't it?

Comment (vpodorozh) [reply app/code/Magento/Catalog/Pricing/Render/FinalPriceBox.php]: Yes, I’ve missed it. This method from Catalog module should always return true.
Will change it tomorrow.

In general this method was added to make it possible to skip salable check in child classes without need to rewrite _toHtml method.

Comment (vpodorozh) [reply app/code/Magento/Catalog/Pricing/Render/FinalPriceBox.php]: @orlangur - I've removed unneeded usage of the variable ( check this https://github.com/magento/magento2/pull/18739/files/f4ec67e5aada173a27586131598d5b24957472d6#diff-5f0a1c702f3ad19b38e301107e2852daR97 ). However, there was a need to add `@SuppressWarnings(PHPMD.UnusedFormalParameter)` phpdoc for `isApplySalableCheck` method to make it pass static tests.

Goal of this method `isApplySalableCheck`: make price rendering depending not only on `isSalable` check but also on addaitional check that can be modified in child classes without need to rewrite [_toHtml ](https://github.com/vpodorozh/magento2/blob/207e2e14306ff1ebea3294d82ed47b8804778d73/app/code/Magento/Catalog/Pricing/Render/FinalPriceBox.php#L66)method.
I've used this "newly added" posibility in this class [Magento\ConfigurableProduct\Pricing\Render\FinalPriceBox ](https://github.com/vpodorozh/magento2/blob/207e2e14306ff1ebea3294d82ed47b8804778d73/app/code/Magento/ConfigurableProduct/Pricing/Render/FinalPriceBox.php#L95)

An other way to achieve this goal is to rewrite _toHtml method that will require a big amount of copy&paste code due to `private` properties and methods usage.

So I do prefer currently implemented solution.

**Comment by magento-engcom-team** [comment]
Hi @vpodorozh. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me $VERSION instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=18739) <br/>All committers have signed the CLA.

**Comment by vpodorozh** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @vpodorozh. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @vpodorozh, here is your [new Magento instance](http://ec2-34-228-235-121.compute-1.amazonaws.com/pr-18739/).
Admin access: [http://ec2-34-228-235-121.compute-1.amazonaws.com/pr-18739//admin](http://ec2-34-228-235-121.compute-1.amazonaws.com/pr-18739//admin)
Login: `admin` Password: `123123q`

**Review by novikor — APPROVED** [review]

**Comment by vpodorozh** [comment]
@orlangur - could you please review my PR?

**Comment by VladimirZaets** [comment]
Hi @vpodorozh. The branch `2.2-develop` will be closed for contribution from **July 15, 2019**. Please finish all your Pull Requests till this date, otherwise, they will be closed.
Thanks for collaboration.

**Comment by orlangur** [comment]
Closing as corresponding 2.3 PR is closed.

**Pull request closed without merging**